## 🌸 Classification Model: Iris Dataset with Decision Tree

In [8]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

# 1. Datensatz laden
iris = load_iris()
X = iris.data
y = iris.target

# 2. Aufteilen in 60% Training, 40% Test (reproduzierbar)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.4, random_state=42, shuffle=True
)

model = DecisionTreeClassifier(
    max_depth=5,
    min_samples_leaf=5,
    min_impurity_decrease=0.1,
    random_state=42
)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(accuracy)

0.9833333333333333


## 🌸 Model Comparison: KNN vs. Naive Bayes (Iris Dataset)

In [9]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

# 1. Datensatz laden
iris = load_iris()
X = iris.data
y = iris.target

# 2. Aufteilen in 60% Training, 40% Test (reproduzierbar)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.4, random_state=42, shuffle=True
)

# 3. KNN-Modell mit 3 Nachbarn
knn = KNeighborsClassifier(n_neighbors=3)
knn.fit(X_train, y_train)
y_pred_knn = knn.predict(X_test)
acc_knn = accuracy_score(y_test, y_pred_knn)

# 4. Naive Bayes (GaussianNB)
nb = GaussianNB()
nb.fit(X_train, y_train)
y_pred_nb = nb.predict(X_test)
acc_nb = accuracy_score(y_test, y_pred_nb)

# 5. Ergebnisse vergleichen
print("Genauigkeit KNN (k=3): {:.2f}%".format(acc_knn * 100))
print("Genauigkeit Naive Bayes (GaussianNB): {:.2f}%".format(acc_nb * 100))


Genauigkeit KNN (k=3): 98.33%
Genauigkeit Naive Bayes (GaussianNB): 96.67%


In [10]:

accuracy = accuracy_score(y_test, y_pred)

print("Genauigkeit: {:.2f}%".format(accuracy * 100))


Genauigkeit: 98.33%


### 🔍 Prediction Analysis: Actual vs Predicted Values

In [11]:
y_pred = knn.predict(X_test)
print("Vorhersagen für die Testdaten:")
print(y_pred)

print("\nTatsächliche Werte:")
print(y_test)

Vorhersagen für die Testdaten:
[1 0 2 1 1 0 1 2 1 1 2 0 0 0 0 1 2 1 1 2 0 2 0 2 2 2 2 2 0 0 0 0 1 0 0 2 1
 0 0 0 2 1 1 0 0 1 1 2 1 2 1 2 1 0 2 1 0 0 0 1]

Tatsächliche Werte:
[1 0 2 1 1 0 1 2 1 1 2 0 0 0 0 1 2 1 1 2 0 2 0 2 2 2 2 2 0 0 0 0 1 0 0 2 1
 0 0 0 2 1 1 0 0 1 2 2 1 2 1 2 1 0 2 1 0 0 0 1]


### 🤖 Model Training & Evaluation: KNN (k=3)

In [12]:

knn = KNeighborsClassifier(n_neighbors=3)
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)
print("Genauigkeit des Modells: {:.2f}%".format(accuracy * 100))

Accuracy: 0.9833333333333333
Genauigkeit des Modells: 98.33%


### 🧠 Interpretation

- Das KNN-Modell erreicht eine Genauigkeit von 98.33%, während Naive Bayes eine Genauigkeit von 96.67% erzielt.
- Beide Modelle performen sehr gut auf dem Iris-Datensatz, was darauf hindeutet, dass die Klassen gut voneinander trennbar sind.
- KNN liefert in diesem Fall leicht bessere Ergebnisse, da es auf der Distanz zwischen Datenpunkten basiert und lokale Strukturen gut erfasst.
- Naive Bayes trifft vereinfachende Annahmen (Unabhängigkeit der Merkmale), was zu minimal geringerer Genauigkeit führen kann.
- Insgesamt zeigt das Ergebnis, dass beide Modelle für dieses Problem gut geeignet sind.